docs/proposals/config_data.md

```diff
@@ -0,0 +1,297 @@
+
+Any long-running system has configuration that is mutated over time.  Changes made to configuration
+data must be made visible to pods consuming data in volumes so that they can respond to those
+changes.
+
+The `resourceVersion` of the `ConfigData` object will be updated by the API server every time the
+object is modified.  After an update, modifications will be made visible to the consumer container:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	Data map[string]string `json:"data,omitempty"`
+}
+
+type ConfigDataList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+
+	Items []ConfigData `json:"items,omitempty"`
+}
+```
+
+A `Registry` implementation for `ConfigData` will be added to `pkg/registry/configdata`.
+
+### Environment variables
+
+The `EnvVarSource` will be extended with a new selector for config data:
+
+```go
+package api
+
+type EnvVarSource struct {
+  ConfigData *ConfigDataSelector `json:"configData,omitempty"`
+}
+
+type ConfigDataSelector struct {
+  APIVersion     string `json:"apiVersion,omitempty"`
+  ConfigDataName string `json:"configDataName"`
+  Key            string `json:"key"`
+}
+```
+
+### Volume Source
+
+A new `ConfigDataVolumeSource` type of volume source containing the `ConfigData` object will be
+added to the `VolumeSource` struct in the API:
+
+```go
+package api
+
+type VolumeSource struct {
+  // other fields omitted
+  ConfigData *ConfigDataVolumeSource `json:"configData,omitempty"`
+}
+
+type ConfigDataVolumeSource struct {
+  Items []ConfigDataVolumeFile `json:"items,omitempty"`
+}
+
+type ConfigDataVolumeFile struct {
+  Path       string             `json:"path"`
```

Comment:
@bgrant0607 this is the path relative tot the volume dir.  same semantaics as equivalent type for downward api volume plugin